### PR TITLE
Fix exception thrown from backlinks field

### LINKED ISSUE #4499: Accessing backlinks field should throw IllegalState instead of IllegalArgument if the object is deleted or not loaded
accessing backlinks of deleted or not loaded object is now throwing `IllegalArgumentException`.

This should be `IllegalStateException`.

### PR DESCRIPTION
**Pull request by zaki50**
fixes #4499

- [x] add test case that reproduce the issue
- [x] fix the code
- [x] change log

### CODE REVIEW

**Comment by zaki50** [comment]
ready for review @realm/java

File: realm/realm-library/src/androidTest/java/io/realm/entities/BacklinksSource.java [hunk 1]
```diff
@@ -18,6 +18,9 @@
 import io.realm.RealmObject;
 
 public class BacklinksSource extends RealmObject {
+    public static final String CLASS_NAME = "BacklinksSource";
```
Comment (dalinaum) [inline realm/realm-library/src/androidTest/java/io/realm/entities/BacklinksSource.java]: Are `CLASS_NAME` and `FIELD_CHILD` used?

File: realm/realm-library/src/androidTest/java/io/realm/entities/BacklinksTarget.java [hunk 2]
```diff
@@ -20,6 +20,10 @@
 import io.realm.annotations.LinkingObjects;
 
 public class BacklinksTarget extends RealmObject {
+    public static final String CLASS_NAME = "BacklinksTarget";
```
Comment (dalinaum) [inline realm/realm-library/src/androidTest/java/io/realm/entities/BacklinksTarget.java]: Same?

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java [hunk 3]
```diff
@@ -401,6 +402,105 @@ public void json_updateList() {
...
 
+    @Test
+    @RunTestInLooperThread
+    public void linkingObjects_IllegalStateException_ifNotYetLoaded() {
+        final Realm realm = looperThread.realm;
+
+        realm.executeTransaction(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                final BacklinksTarget target1 = realm.createObject(BacklinksTarget.class);
```
Comment (dalinaum) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java]: Is there other target variable?

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java [hunk 4]
```diff
@@ -401,6 +402,105 @@ public void json_updateList() {
...
+
+    @Test
+    @RunTestInLooperThread
+    public void linkingObjects_IllegalStateException_ifDeleted() {
+        final Realm realm = looperThread.realm;
+
+        realm.executeTransaction(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                final BacklinksTarget target1 = realm.createObject(BacklinksTarget.class);
```
Comment (dalinaum) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java]: Same

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java [hunk 5]
```diff
@@ -401,6 +402,105 @@ public void json_updateList() {
...
+                target1.setId(1);
+
+                final BacklinksSource source = realm.createObject(BacklinksSource.class);
+                source.setChild(target1);
+            }
+        });
+
+        final BacklinksTarget target = realm.where(BacklinksTarget.class)
```
Comment (dalinaum) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java]: Oh. I just saw this line. `target1` and `target` are little weird. Maybe there are better names.

File: realm/realm-library/src/androidTest/java/io/realm/entities/BacklinksSource.java (hunk 1 above)
Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/entities/BacklinksSource.java]: Not used now, but will be used in https://github.com/realm/realm-java/pull/4492

File: realm/realm-library/src/main/java/io/realm/internal/UncheckedRow.java [hunk 6]
```diff
@@ -188,6 +188,13 @@ public LinkView getLinkList(long columnIndex) {
         return new LinkView(context, parent, columnIndex, nativeLinkViewPtr);
     }
 
+    @Override
+    public void checkIfBacklinkAvailable() {
```
Comment (bmeike) [inline realm/realm-library/src/main/java/io/realm/internal/UncheckedRow.java]: This function has the wrong name.  It is not checking to see if backlinks are available.  It is checking to see if the object is attached.

**Comment by bmeike** [comment]
I would prefer that you change the name of the new method, to reflect its actual function.  Otherwise, LGTM.

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java (hunk 5 above)
Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java]: Both are the same object, but one is typed, another is dynamic.
In this test case, we don't need to use different variable name for them.
I'll update them to use the same variable name.

File: realm/realm-library/src/main/java/io/realm/internal/UncheckedRow.java (hunk 6 above)
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/internal/UncheckedRow.java]: The main concern now is whether we can provide a backlinks object to users, but yes, that's the same with whether it's attached or not. And latter is more versatile. 

How about `checkIfAttached()`?

**Comment by zaki50** [comment]
@bmeike Is `Row.checkIfAttached()` fine?
If yes, I'll merge this PR.

**Comment by zaki50** [comment]
@bmeike To progress #4492, I'm merging this PR.
If there is a better name for `Row.checkIfAttached()`, I'll update it in another PR.

**Pull request merged**